spinalcordtoolbox/deepseg/core.py

```diff
@@ -0,0 +1,144 @@
+# coding: utf-8
+"""
+Interface API for the deepseg module, which performs segmentation using deep learning with the ivadomed package.
+"""
+
+
+import logging
+import numpy as np
+import nibabel as nib
+import ivadomed as imed
+import ivadomed.utils
+import ivadomed.postprocessing
+
+import spinalcordtoolbox as sct
+import spinalcordtoolbox.deepseg.models
+
+
+logger = logging.getLogger(__name__)
+
+# Default values if not asked during CLI call and if not present in json metadata.
+DEFAULT_THRESHOLD = 0.9
+DEFAULT_KEEP_LARGEST_OBJECT = True
+DEFAULT_FILL_HOLES = True
+
+
+class ParamDeepseg:
+    """
+    Parameters for deepseg module.
+    """
+    def __init__(self):
+        self.threshold = None
+        self.keep_largest_object = None
+        self.fill_holes = None
+        self.output_suffix = '_seg'
+        self.remove_temp_files = 1
+        self.verbose = 1
+        # TODO: add threshold, keep_big_obj
+
+
+class PostProcessing:
+    """
+    Deals with post-processing of the segmentation. Consider param (i.e. user's flag) with more priority than
+    metadata (i.e. from model's json file).
+    """
+    def __init__(self, param, metadata):
+        """
+        :param param: class ParamDeepseg: Defined by user's parameter
+        :param metadata: dict: From model's json metadata
+        """
+        self.param = param
+        self.metadata = metadata
+
+    def threshold(self, nii_seg):
+        """
+        Threshold the prediction. For no prediction, set 'threshold' to 0.
+        """
+        if self.param.threshold:
+            thr = self.param.threshold
+        else:
+            if 'threshold' in self.metadata:
+                thr = self.metadata.threshold
+            else:
+                logger.warning("'threshold' is not defined in the model json file. Using threshold of: {}".format(
+                    DEFAULT_THRESHOLD))
+                thr = DEFAULT_THRESHOLD
+        if thr:
+            nii_seg = imed.postprocessing.threshold_predictions(nii_seg, thr)
+        return nii_seg
+
+    def keep_largest_object(self, nii_seg):
+        """
+        Only keep largest object
+        """
+        # TODO: This if/elif below is ugly. Cannot think of something better for now...
+        do_process = DEFAULT_KEEP_LARGEST_OBJECT
+        if self.param.keep_largest_object is True:
+            do_process = True
+        elif self.param.keep_largest_object is None:
+            if 'keep_largest_object' in self.metadata:
+                do_process = self.metadata.keep_largest_object
+        if do_process:
```

Comment:
I thought about what possible values everything could take here. It looks like you want `self.param` taking precedence, then `self.metadata`, then falling back to `DEFAULT_`, with `None` or `not in ...` meaning "undefined, so try the next option".

I would say you want `do_process = param or metadata or default`, except that will misinterpret `False` as "see the next" instead of "stop here".

 I rewrote this expression as a function I could test exhaustively with all the possible values; I made one simplification that I don't think would change the meaning -- to represent `'keep_largest_object' not in self.metadata` as `metadata is None`.

```
from pprint import pprint
def do_process(param, metadata, default):
    if param is True:
        return True
    elif param is None:
        if metadata is not None: # rewriting a: 'keep_largest_object' not in self.metadata <=> metadata is None
            return metadata
    return default

pprint([(param, metadata, default, "=>", do_process(param, metadata, default)) for param in [True, False, None] for metadata in [True, False, None] for default in [True, False]])
```

this gave me this truth table:

```
[(True, True, True, '=>', True),
 (True, True, False, '=>', True),
 (True, False, True, '=>', True),
 (True, False, False, '=>', True),
 (True, None, True, '=>', True),
 (True, None, False, '=>', True),
 (False, True, True, '=>', True),
 (False, True, False, '=>', False),
 (False, False, True, '=>', True),
 (False, False, False, '=>', False),
 (False, None, True, '=>', True),
 (False, None, False, '=>', False),
 (None, True, True, '=>', True),
 (None, True, False, '=>', True),
 (None, False, True, '=>', False),
 (None, False, False, '=>', False),
 (None, None, True, '=>', True),
 (None, None, False, '=>', False)]
```

In all cases where `param` is True, it's True (of course, that's written right into the logic). But when it's False it skips `metadata` and uses `default` instead.

Is that what you really meant? `param=False` is the same as `param=None`, but `metadata=False` means "stop"? I think that's weird. I think the actual logic you want is

```
def do_process(param, metadata, default):
    if param is True:
        return True
    elif param is False:
        return False
    elif param is None:
        if metadata is not None:
            return metadata
    return default
```

which can be written shorter as

```
def do_process(param, metadata, default):
    if param is not None:
        return param
    if metadata is not None:
        return metadata
    return default
```

The truth table for this is

```
[(True, True, True, '=>', True),
 (True, True, False, '=>', True),
 (True, False, True, '=>', True),
 (True, False, False, '=>', True),
 (True, None, True, '=>', True),
 (True, None, False, '=>', True),
 (False, True, True, '=>', False),
 (False, True, False, '=>', False),
 (False, False, True, '=>', False),
 (False, False, False, '=>', False),
 (False, None, True, '=>', False),
 (False, None, False, '=>', False),
 (None, True, True, '=>', True),
 (None, True, False, '=>', True),
 (None, False, True, '=>', False),
 (None, False, False, '=>', False),
 (None, None, True, '=>', True),
 (None, None, False, '=>', False)]
```

Reply:
The other tricky thing I see is that 'undefined' is represented in two different ways: for `self.param`, it's represented by `None`, for `self.metadata` it's represented by an actual missing value. What should happen if someone set `self.metadata['keep_largest_object'] = None`? Or if someone did `del self.param.keep_largest_object`? If you could change the representation of options this would be easier. If you can't, we're going to have to accept some baseline ugliness.

Anyway, if you accept my changing of your logic above and the need to deal with different representations, you can rewrite this with the ternary hook:

```suggestion
        if (self.param.keep_largest_object
            if self.param.keep_largest_object is not None
            else self.metadata['keep_largest_object']
            if 'keep_largest_object' in self.metadata
            else DEFAULT_KEEP_LARGEST_OBJECT):
```

Python has a shortcut for the `if .. in .. else` part, `dict.get(...[, default])`:

```suggestion
        if (self.param.keep_largest_object
            if self.param.keep_largest_object is not None
            else self.metadata.get('keep_largest_object',
                    DEFAULT_KEEP_LARGEST_OBJECT):
```

Reply:
You could also map None => missing directly, and then make use of `dict.get()`:

```suggestion
        if ({k: v for k,v in self.param.__dict__.items() if v is not None}).get('keep_largest_object', self.metadata.get('keep_largest_object', DEFAULT_KEEP_LARGEST_OBJECT)):
```

Reply:
Or what about doing:

```
DEFAULTS = {'threshold': 0.9, 'keep_largest_object': True, 'fill_holes': True}

class PostProcessing:
    """
    Deals with post-processing of the segmentation. Consider param (i.e. user's flag) with more priority than
    metadata (i.e. from model's json file).
    """
    def __init__(self, param, metadata):
        """
        :param param: class ParamDeepseg: Defined by user's parameter
        :param metadata: dict: From model's json metadata
        """

        # param overrides metadata overrides defaults
        self.param = dict(DEFAULTS)
        self.param.update(metadata)
        self.param.update({k: v for k,v in param.__dict__.items() if v is not None})

        [...]

    def keep_largest_object(self, nii_seg):
        """
        Only keep largest object
        """
        if self.param['keep_largest_object']:
            [...]
```

Reply:
If you can drop `class ParamDeepseg` and just have callers pass a dict with the settings they want to override then you don't even need the extra inner dict-comprehension.

Reply:
merged strategy implemented in bea7a3cb50862cfebe64d36dc71647cf4a7ae1dc